pkg/api/dirs_test.go

```diff
@@ -0,0 +1,247 @@
+			Message: "could not store dir",
+			Code:    http.StatusInternalServerError,
+		}, nil)
+	})
+
+	t.Run("non tar file", func(t *testing.T) {
+		f := writeAndOpenFile(t, "non-tar-file", []byte("some data"))
+		defer os.Remove(f.Name())
+		defer f.Close()
+
+		jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, f, http.StatusInternalServerError, jsonhttp.StatusResponse{
+			Message: "could not store dir",
+			Code:    http.StatusInternalServerError,
+		}, nil)
+	})
+
+	for _, tc := range []dirTestCase{
+		{
+			expectedHash: "2fa041bd35ebff676727eb3023272f43b1e0fa71c8735cc1a7487e9131f963c4",
+			files: []dirTestCaseFile{
+				{
+					data:      []byte("first file data"),
+					name:      "file1",
+					path:      "",
+					reference: swarm.MustParseHexAddress("3c07cd2cf5c46208d69d554b038f4dce203f53ac02cb8a313a0fe1e3fe6cc3cf"),
+					headers: http.Header{
+						"Content-Type": {""},
+					},
+				},
+				{
+					data:      []byte("second file data"),
+					name:      "file2",
+					path:      "",
+					reference: swarm.MustParseHexAddress("47e1a2a8f16e02da187fac791d57e6794f3e9b5d2400edd00235da749ad36683"),
+					headers: http.Header{
+						"Content-Type": {""},
+					},
+				},
+			},
+		},
+		{
+			expectedHash: "c3cb9fbe2efa7bbc979245d9bac1400bd4894371776b7560309d49e687514dd6",
+			files: []dirTestCaseFile{
+				{
+					data:      []byte("robots text"),
+					name:      "robots.txt",
+					path:      "",
+					reference: swarm.MustParseHexAddress("17b96d0a800edca59aaf7e40c6053f7c4c0fb80dd2eb3f8663d51876bf350b12"),
+					headers: http.Header{
+						"Content-Type": {"text/plain; charset=utf-8"},
+					},
+				},
+				{
+					data:      []byte("image 1"),
+					name:      "1.png",
+					path:      "img",
+					reference: swarm.MustParseHexAddress("3c1b3fc640e67f0595d9c1db23f10c7a2b0bdc9843b0e27c53e2ac2a2d6c4674"),
+					headers: http.Header{
+						"Content-Type": {"image/png"},
+					},
+				},
+				{
+					data:      []byte("image 2"),
+					name:      "2.png",
+					path:      "img",
+					reference: swarm.MustParseHexAddress("b234ea7954cab7b2ccc5e07fe8487e932df11b2275db6b55afcbb7bad0be73fb"),
+					headers: http.Header{
+						"Content-Type": {"image/png"},
+					},
+				},
+			},
+		},
+	} {
+		t.Run("valid tar", func(t *testing.T) {
+			// create and collect all files in the test case
+			dirFiles := make([]*os.File, len(tc.files))
+			for i, file := range tc.files {
+				// create dir if the file is nested
+				if file.path != "" {
+					if _, err := os.Stat(file.path); os.IsNotExist(err) {
+						if err := os.Mkdir(file.path, 0700); err != nil {
+							t.Fatal(err)
+						}
+						defer os.RemoveAll(file.path)
+					}
+				}
+				f := writeAndOpenFile(t, path.Join(file.path, file.name), file.data)
+				defer os.Remove(f.Name())
+				defer f.Close()
```

Comment:
not all that critical but the `defer` will only execute after the entire test case is finished. in general i would avoid usage of `defer` statements inside `for` loops as it is misleading to think the execute after the for loop is done, which is not the case

Reply:
where would you put these statements instead?

it cannot be before line `170`, and we'd have to keep track of every file and dir created in the test

Reply:
you could create the tar and delete the files in the same function, the indirection you're using here is a bit crooked since you're opening file handles, keeping them open and collecting them for some other function to use. you could create the files one by one and just add them sequentially to the tarball and delete them. also, it might be way simpler to just use a test fixture here and have a prefabricated small tarball in the codebase in a testdata subdirectory. just load that up, unpack, verify, tschuess, instead of all of the constant regeneration of the same data. increases predictability too

Reply:
see old swarm codebase, `migration_test.go#TestMigrateSanctuaryFixture`

Reply:
thanks for all this thoughtful input :heart: 

so i've opted to continue to generate the tar buffer, but without the interaction to disk. as a result, the code is much simpler now.

please see [this reply above](https://github.com/ethersphere/bee/pull/447#discussion_r461297210) and lmk what you think